exonum/src/node/mod.rs

```diff
@@ -547,37 +543,55 @@ impl NodeHandler {
 
     /// Sends the given message to a peer by its public key.
     pub fn send_to_peer(&mut self, public_key: PublicKey, message: &RawMessage) {
-        if let Some(conn) = self.state.peers().get(&public_key) {
-            let address = conn.addr();
-            trace!("Send to address: {}", address);
-            let request = NetworkRequest::SendMessage(address, message.clone());
-            self.channel.network_requests.send(request).log_error();
-        } else {
-            warn!("Hasn't connection with peer {:?}", public_key);
+        let address = {
+            if let Some(conn) = self.state.peers().get(&public_key) {
+                Some(conn.addr())
+            } else {
+                warn!("Hasn't connection with peer {:?}", public_key);
+                None
+            }
+        };
+
+        if let Some(address) = address {
+            self.send_to_addr(&address, message);
         }
     }
 
     /// Sends `RawMessage` to the specified address.
     pub fn send_to_addr(&mut self, address: &SocketAddr, message: &RawMessage) {
-        trace!("Send to address: {}", address);
-        let request = NetworkRequest::SendMessage(*address, message.clone());
-        self.channel.network_requests.send(request).log_error();
+        let public_key = self.state.connect_list().find_key_by_address(&address);
+
+        match public_key {
+            Some(public_key) => {
+                trace!("Send to address: {}", address);
+                let request = NetworkRequest::SendMessage(*address, message.clone(), *public_key);
+                self.channel.network_requests.send(request).log_error();
+            }
+            _ => {
+                warn!("Peer is not in the ConnectList {:?}", public_key);
+            }
+        }
     }
 
     /// Broadcasts given message to all peers.
     pub fn broadcast(&mut self, message: &RawMessage) {
-        for conn in self.state.peers().values() {
+        let peers: Vec<Connect> = { self.state.peers().values().cloned().collect() };
+
+        for conn in peers {
             let address = conn.addr();
-            trace!("Send to address: {}", address);
-            let request = NetworkRequest::SendMessage(address, message.clone());
-            self.channel.network_requests.send(request).log_error();
+            self.send_to_addr(&address, message);
         }
     }
 
     /// Performs connection to the specified network address.
     pub fn connect(&mut self, address: &SocketAddr) {
         let connect = self.state.our_connect_message().clone();
-        self.send_to_addr(address, connect.raw());
+
+        if self.state.connect_list().address_allowed(&address) {
+            self.send_to_addr(address, connect.raw());
+        } else {
+            warn!("Peer {:?} is not in the ConnectList!", address);
```

Comment:
I'd suggest `Attempt to connect to the peer with public key {:?} which is not in the ConnectList`.